nix/ext/pg_cron.nix

```diff
@@ -1,31 +1,121 @@
-  version = "1.6.4";
+let
+  allVersions = {
+    "1.3.1" = {
+      rev = "v1.3.1";
+      hash = "sha256-rXotNOtQNmA55ErNxGoNSKZ0pP1uxEVlDGITFHuqGG4=";
+      postPatch = ''
+        # Add necessary includes
+        substituteInPlace src/pg_cron.c \
+          --replace '#include "postgres.h"' '#include "postgres.h"
+      #include "commands/async.h"
+      #include "miscadmin.h"'
 
-  buildInputs = [ postgresql ];
+        # Update function calls to use PostgreSQL 15 APIs
+        substituteInPlace src/pg_cron.c \
+          --replace 'ProcessCompletedNotifies();' '/* ProcessCompletedNotifies removed */' \
+          --replace 'pg_analyze_and_rewrite(parsetree, sql, NULL, 0,NULL);' 'pg_analyze_and_rewrite_fixedparams(parsetree, sql, NULL, 0, NULL);'
+      '';
+    };
+    "1.4.2" = {
+      rev = "v1.4.2";
+      hash = "sha256-P0Fd10Q1p+KrExb35G6otHpc6pD61WnMll45H2jkevM=";
+    };
+    "1.6.4" = {
+      rev = "v1.6.4";
+      hash = "sha256-t1DpFkPiSfdoGG2NgNT7g1lkvSooZoRoUrix6cBID40=";
+    };
+    "1.5.2" = {
+      rev = "v1.5.2";
+      hash = "sha256-+quVWbKJy6wXpL/zwTk5FF7sYwHA7I97WhWmPO/HSZ4=";
+    };
+  };
+
+  mkPgCron = pgCronVersion: { rev, hash, postPatch ? "" }: stdenv.mkDerivation {
+    pname = "pg_cron";
+    version = "${pgCronVersion}-pg${lib.versions.major postgresql.version}";
+
+    buildInputs = [ postgresql ];
+    inherit postPatch;
+
+    src = fetchFromGitHub {
+      owner = "citusdata";
+      repo = "pg_cron";
+      inherit rev hash;
+    };
+
+    buildPhase = ''
+      make PG_CONFIG=${postgresql}/bin/pg_config
+  
+      # Create version-specific SQL file
+      cp pg_cron.sql pg_cron--${pgCronVersion}.sql
+
+      # Create versioned control file with modified module path
+      sed -e "/^default_version =/d" \
+          -e "s|^module_pathname = .*|module_pathname = '\$libdir/pg_cron'|" \
+          pg_cron.control > pg_cron--${pgCronVersion}.control
+    '';
 
-  src = fetchFromGitHub {
-    owner  = "citusdata";
-    repo   = pname;
-    rev    = "v${version}";
-    hash = "sha256-t1DpFkPiSfdoGG2NgNT7g1lkvSooZoRoUrix6cBID40=";
+    installPhase = ''
+      mkdir -p $out/{lib,share/postgresql/extension}
+      
+      # Install versioned library
+      install -Dm755 pg_cron${postgresql.dlSuffix} $out/lib/pg_cron-${pgCronVersion}${postgresql.dlSuffix}
+      
+      # Install version-specific files
+      install -Dm644 pg_cron--${pgCronVersion}.sql $out/share/postgresql/extension/
+      install -Dm644 pg_cron--${pgCronVersion}.control $out/share/postgresql/extension/
+      
+      # Install upgrade scripts
+      find . -name 'pg_cron--*--*.sql' -exec install -Dm644 {} $out/share/postgresql/extension/ \;
+    '';
   };
 
+  getVersions = pg:
+    if lib.versionAtLeast pg.version "17"
+    then { "1.6.4" = allVersions."1.6.4"; }
+    else allVersions;
+
+  allVersionsForPg = lib.mapAttrs mkPgCron (getVersions postgresql);
+
+in
+stdenv.mkDerivation {
+  pname = "pg_cron-all";
+  version = "multi";
```

Comment:
Q: Could this `version` give us trouble later on since it looks like it's not meant to be modified? Say if we modify the individual pg cron versions, could the hardcoded `multi` version cause a "permanent stale cache"?

Reply:
@steve-chavez it's a good call, I will create an increment-able version there, that we can advance when/if we change the the package at all.

Reply:
@samrose Not sure if possible, but maybe this version can be a concatenation of all the individual pg_cron versions. That way it doesn't have to be updated manually.

Reply:
@steve-chavez hmm yeah it probably is possible. Ok!

Reply:
Done!